handlers/connection.go

```diff
@@ -0,0 +1,168 @@
+	DestinationType  string      `json:"type"`
+	ConnectionConfig interface{} `json:"config"`
+}
+
+type ConnectionTestHandler struct {
+}
+
+type SnowflakeExternalConfig struct {
+	SnowflakeConfig adapters.SnowflakeConfig `json:"snowflake"`
+	S3Config        adapters.S3Config        `json:"s3"`
+}
+
+func testConnection(config ConnectionConfig) error {
+	switch config.DestinationType {
+	case "postgres":
+		var postgresConfig adapters.DataSourceConfig
+		body, err := json.Marshal(config.ConnectionConfig)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(body, &postgresConfig)
+		if err != nil {
+			return err
+		}
+		if err := postgresConfig.Validate(); err != nil {
+			return err
+		}
+		postgres, err := adapters.NewPostgres(context.Background(), &postgresConfig)
+		if err != nil {
+			return err
+		}
+		defer postgres.Close()
+		return postgres.Test()
+
+	case "clickhouse":
+		var chConfig adapters.ClickHouseConfig
+		body, err := json.Marshal(config.ConnectionConfig)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(body, &chConfig)
+		if err != nil {
+			return err
+		}
+		if err = chConfig.Validate(); err != nil {
+			return err
+		}
+		tableStatementFactory, err := adapters.NewTableStatementFactory(&chConfig)
+		if err != nil {
+			return err
+		}
+		nonNullFields := map[string]bool{"eventn_ctx_event_id": true, "_timestamp": true}
+		for i := range chConfig.Dsns {
+			var resultError error
+			resultError = nil
+			ch, err := adapters.NewClickHouse(context.Background(), chConfig.Dsns[i], chConfig.Database, chConfig.Cluster, chConfig.Tls, tableStatementFactory, nonNullFields)
+			if err != nil {
+				resultError = err
+				continue
+			}
+			resultError = ch.Test()
+			if err = ch.Close(); err != nil {
+				logging.Warn("Failed to close clickhouse datasource %s", err)
+			}
+			if resultError == nil {
+				return nil
+			}
+		}
+		return nil
+
+	case "redshift":
+		var rsConfig adapters.RedshiftConfig
+		body, err := json.Marshal(config.ConnectionConfig)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(body, &rsConfig)
+		if err != nil {
+			return err
+		}
+		if err = rsConfig.DbConfig.Validate(); err != nil {
+			return err
+		}
+		if rsConfig.S3Config != nil {
+			if err = rsConfig.S3Config.Validate(); err != nil {
+				return err
+			}
+		}
+		redshift, err := adapters.NewAwsRedshift(context.Background(), rsConfig.DbConfig, rsConfig.S3Config)
+		if err != nil {
+			return err
+		}
+		defer redshift.Close()
+		return redshift.Test()
+	case "bigquery":
+		return errors.New("bigquery validation is not supported")
+	case "s3":
+		s3Config := adapters.S3Config{}
+		body, err := json.Marshal(config.ConnectionConfig)
+		if err != nil {
+			return err
+		}
+		if err = json.Unmarshal(body, &s3Config); err != nil {
+			return err
+		}
+		if err = s3Config.Validate(); err != nil {
+			return err
+		}
+		s3, err := adapters.NewS3(&s3Config)
+		defer s3.Close()
+		return err
+	case "snowflake":
```

Comment:
Since we've supported only postgres,clickhouse and redshift, we can omit snowflake and s3 test connections, because we don't know config structure (and snowflake can be configured via s3,gcp)